bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py

```diff
@@ -0,0 +1,509 @@
+        '''
+
+        # TODO
+        # - Do a more thorough type conversion
+        # - Use RDF helper functions to convert jitted strings to lambdas
+
+        # Argument type conversion
+        for narg, arg in enumerate(operation.args):
+            if (narg > 0):
+                self.graph_nodes += ', '
+
+            if isinstance(arg, str):
+                self.graph_nodes += '"{}"'.format(arg)
+            elif isinstance(arg, tuple):
+                self.graph_nodes += '{'
+                for nelem, elem in enumerate(arg):
+                    if nelem > 0:
+                        self.graph_nodes += ','
+                    if isinstance(elem, str):
+                        self.graph_nodes += '"{}"'.format(elem)
+                    else:
+                        self.graph_nodes += '{}'.format(elem)
+                self.graph_nodes += '}'
+
+        # Make Snapshot lazy
+        # TODO: Do a proper processing of the args (user might have specified
+        # her own options object)
+        if operation.name == 'Snapshot':
+            self.graph_nodes += ', "", lazy_options'
+
+
+    def execute(self, rdf):
+        '''
+        Compiles the workflow generation code and executes it.
+
+        Args:
+            rdf (ROOT::RDF::RNode): object that represents the dataset on
+                which to execute the workflow.
+
+        Returns:
+            tuple: the first element is the list of results of the actions in
+                the C++ workflow, the second element is the list of result types
+                corresponding to those actions.
+        '''
+
+        function_id = self._compile()
+        return self._run_function(rdf, function_id)
+
+    def _compile(self):
+        '''
+        Generates the workflow code C++ file and compiles it with ACLiC
+        into a shared library. The library is also loaded as part of the
+        `TSystem::CompileMacro` call.
+
+        The name of the generated C++ file contains both a hash of its
+        code and the ID of the process that created it. This is done to
+        prevent clashes between multiple (non-sandboxed) worker processes
+        that try to write to the same file concurrently.
+
+        A class-level cache keeps track of the workflows that have been already
+        compiled to prevent unncessary recompilation (e.g. when a worker
+        process runs multiple times the same workflow).
+
+        Returns:
+            string: the id of the function to be executed. Such id is appended
+                to CppWorkflow.FUNCTION_NAME to prevent name clashes (a worker
+                process might compile and load multiple workflow functions).
+                A function id is mostly the hash of its code.
+        '''
+
+        # TODO: Make this function thread-safe? To support Dask threaded
+        # workers
+
+        code = self._get_code()
+        code_hash = hash(code)
+        function_id = str(code_hash).replace('-', '_', 1)
```

Comment:
A hash can be a negative integer value (thus with a `-` in its string representation). I replace `-` by `_` so that the hash can be used as part of the workflow generator function name (`-` is not allowed in function names).

It's not obvious, I added a comment to explain this :)

Reply:
Alright nice thanks!